src/cljs/s3_beam/client.cljs

```diff
@@ -31,28 +65,80 @@
       (.append fd (name k) v))
     fd))
 
-(defn upload-file [upload-info ch]
-  (let [sig-fields [:key :Content-Type :success_action_status :policy :AWSAccessKeyId :signature :acl]
-        signature  (select-keys (:signature upload-info) sig-fields)
-        form-data  (formdata-from-map (merge signature {:file (:f upload-info)}))]
-    (xhr/send
-     (:action (:signature upload-info))
-     (fn [res]
-       (let [loc (aget (.getElementsByTagName (.getResponseXml (.-target res)) "Location") 0)]
-         (put! ch (gdom/getTextContent loc))
-         (close! ch)))
-     "POST"
-     form-data)))
+(defn xml->map
+  "Transforms a shallow XML document into a map."
+  [xml]
+  (let [kv (->> (array-seq (.. xml -firstChild -childNodes))
+                (map (fn [n] [(.toLowerCase (.-nodeName n))
+                             (gdom/getTextContent n)])))]
+    (zipmap (map (comp keyword first) kv) (map second kv))))
+
+(defn upload-file
+  "Take the signature information `upload-info`, and a channel `ch`,
+  and does the S3 upload.
+  Returns a map with:
+    :file - the File that was uploaded
+    :response - a map of S3 metadata: :key, :location, :bucket, :etag
+    :identifier - an identifier pass-through value for caller to identify the file with
+  If the upload-info has an :error-code key because the signing failed, skips the xhr/send of the file,
+  and simply replies with the error messages and codes from the failed upload-info sign attempt."
+  [upload-info ch]
+  (if (contains? upload-info :error-code)
+    (do
+      (put! ch (select-keys upload-info [:identifier :error-code :error-message :http-error-code]))
```

Comment:
Adding functions that don't do anything is almost always making it harder to understand how the code works, imo :)